src/main/java/io/split/android/client/TrackClientImpl.java

```diff
@@ -336,14 +336,20 @@ public void run() {
             } catch (InterruptedException e) {
                 Logger.w("Consumer thread was interrupted. Exiting...");
                 //Saving event due to consumer interruption
-                _storageManager.saveEvents(new EventsChunk(events));
+                _storageManager.saveEvents(new EventsChunk(eventsCopy()));
             }
         }
 
         private List<Event> newEventList() {
             return Collections.synchronizedList(new ArrayList<>());
         }
 
+        private List<Event> eventsCopy() {
+            synchronized (events) {
+                return new ArrayList<>(events);
```

Comment:
maybe doing the events immutable could help with threading issues, even using Kotlin can make this easier.